codalab/worker_manager/slurm_batch_worker_manager.py

```diff
@@ -0,0 +1,295 @@
+            worker_network_prefix,
+            # always set in Slurm worker manager to ensure safe shutdown
+            '--pass-down-termination',
+        ]
+        if self.args.worker_tag:
+            command.extend(['--tag', self.args.worker_tag])
+        if self.args.password_file:
+            command.extend(['--password-file', self.args.password_file])
+
+        slurm_args = self.create_slurm_args(worker_id)
+        job_definition = self.create_job_definition(slurm_args=slurm_args, command=command)
+
+        # Not submit job to Slurm if dry run is specified
+        if self.args.dry_run:
+            return
+
+        batch_script = str(slurm_work_dir.joinpath(slurm_args['job-name'] + '.slurm'))
+        self.save_job_definition(batch_script, job_definition)
+        job_id_str = self.run_command([self.SBATCH, batch_script])
+
+        match = re.match(self.SBATCH_COMMAND_RETRUN_REGEX, job_id_str)
+        if match is not None:
+            job_id = match.group(1)
+        else:
+            logger.error("Cannot find job_id in {}.".format(job_id_str))
+            return
+
+        # Add the newly submitted job to submitted_jobs for tracking purpose
+        self.submitted_jobs.add(job_id)
+
+    def run_command(self, command):
+        proc = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        output, errors = proc.communicate()
+        if output:
+            logger.info("Executed command: {}".format(' '.join(command)))
+            return output.decode()
+        if errors:
+            print("Failed to execute {}: {}: {}".format(' '.join(command), errors, proc.returncode))
+            logger.error(errors)
+        return ""
+
+    def save_job_definition(self, job_file, job_definition):
+        """
+        Save the batch job definition to file.
+        :param job_file: a file storing the Slurm batch job configuration
+        :param job_definition: the job definition of a Slurm batch job
+        :return:
+        """
+        with open(job_file, 'w') as f:
+            f.write(job_definition)
+        logger.info("Saved the Slurm Batch Job Definition to {}".format(job_file))
+
+    def create_job_definition(self, slurm_args, command):
+        """
+        Create a Slurm batch job definition structured as a list of Slurm batch arguments and a srun command
+        :param slurm_args: arguments for launching a Slurm batch job
+        :param command: arguments for starting a CodaLab worker
+        :return: a string containing the Slurm batch job definition
+        """
+        sbatch_args = [
+            '{} --{}={}'.format(self.SBATCH_PREFIX, key, slurm_args[key])
+            for key in sorted(slurm_args.keys())
+        ]
+
+        # Check the existence of environment variables CODALAB_USERNAME and
+        # CODALAB_PASSWORD when password_file is not given.
+        worker_authentication = (
+            '[ ! -z "${CODALAB_USERNAME}" ] && [ ! -z "${CODALAB_PASSWORD}" ] || exit;\n\n'
```

Comment:
Would it be possible to add some sort of logging message that explicitly states that the worker is exiting because there's no `--password-file` and `CODALAB_USERNAME` and `CODALAB_PASSWORD` are not set?